### DISCUSSION TITLE
How to schedule?

### DISCUSSION THREAD

**Discussion by cristianvogel** [djipco/webmidi#213] [Q&A] [answered]
Hello! 

Many thanks for maintaining this excellent library. It is clear, well documented and easy to use.
I am working to add a midi out feature to an app written by someone else which originally sequences sample based audio buffers with a complicated  promise based system. 

I am trying to inject WebMidi in there. 

I have succeeded more or less, but wondered if I would get tighter MIDI events, by scheduling them in a similar way to the original code does with audio. I *think* the original code uses a scheduling list, which I believe WebMidi is capable of for notes. 

But I don't really understand how to do this when my relative reference is another timer, not `WebMidi.time` 

`WebMidi.outputs[0].channels[1].playNote("C3", {time: WebMidi.time + 3000});
WebMidi.outputs[0].channels[1].playNote("C3", {time: "+2000"});`

The code base I am working on his here [Twotone](https://github.com/sonifydata/twotone/tree/v2.1-dev)
and I try to inject the WebMidi note out [here](https://github.com/sonifydata/twotone/blob/v2.1-dev/src/engine/types/scale/index.js) 

_The original author of this program has retired from this project and is unavailable, so I'm trying my best to understand how to add a Midi Out feature into a complicated system of callbacks_

**Comment by djipco** [comment]
I'm not quite clear on what you are asking. One thing that is for sure is that WEBMIDI.js can connect to existing input and output ports but **cannot** create new ports. 

By the way, `WebMidi.time` is simply an alias to `performance.now()`.

**Reply by cristianvogel** [reply]
Thanks for your answer. Focussing on the `WebMidi.time` question. In the codebase I am working with, it seems that   a performance timer starts when the user presses play, like as if performance.now() could be rest to 0?  I guess I am asking how I can make relative scheduling of WebMidi notes against a different highres timer than `performance.now` which I think is not resettable? Or is it?

**Comment by djipco** [comment, marked as answer]
You cannot reset `performance.now()`. However, what you can do is save the offset between the moment you press "play" in your system and `WebMidi.time`. Then, whenever you want to schedule a note, you just have to add this offset. 

This got me thinking though... perhaps this could be added to the library. It wouldn't be hard to implement what I wrote above into the lib so that `WebMidi.time` can be reset. I'll add that to my list of [things to evaluate](https://webmidijs.org/docs/roadmap/under-evaluation).

**Reply by cristianvogel** [reply]
I think thats a good idea. I found this guide by @haszari which might make this feature addition straightforward for you. 

https://github.com/haszari/sync-web-audio-web-midi

**Reply by djipco** [reply]
Awesome! Thanks for the link. It makes total sense to find a sensible way to sync `WebMidi.time` with an external source (such as Web Audio or others).

**Reply by haszari** [reply]
Coolness! 🙌
